cabal-install/src/Distribution/Client/ProjectBuilding.hs

```diff
@@ -652,57 +656,64 @@ rebuildTargets
     | otherwise = do
         -- Concurrency control: create the job controller and concurrency limits
         -- for downloading, building and installing.
-        jobControl <-
-          if isParallelBuild
-            then newParallelJobControl buildSettingNumJobs
-            else newSerialJobControl
+        mkJobControl <- case buildSettingNumJobs of
+          Serial -> newSerialJobControl
+          NumJobs n -> newParallelJobControl (fromMaybe numberOfProcessors n)
+          UseSem n ->
+            if jsemSupported compiler
+              then newSemaphoreJobControl n
+              else do
+                warn verbosity "-jsem is not supported by the selected compiler, falling back to normal parallelism control."
+                newParallelJobControl n
         registerLock <- newLock -- serialise registration
         cacheLock <- newLock -- serialise access to setup exe cache
         -- TODO: [code cleanup] eliminate setup exe cache
-        debug verbosity $
+        info verbosity $
           "Executing install plan "
-            ++ if isParallelBuild
-              then " in parallel using " ++ show buildSettingNumJobs ++ " threads."
-              else " serially."
+            ++ case buildSettingNumJobs of
+              NumJobs n -> " in parallel using " ++ show n ++ " threads."
+              UseSem n -> " in parallel using a semaphore with " ++ show n ++ " slots."
+              Serial -> " serially."
 
         createDirectoryIfMissingVerbose verbosity True distBuildRootDirectory
         createDirectoryIfMissingVerbose verbosity True distTempDirectory
         traverse_ (createPackageDBIfMissing verbosity compiler progdb) packageDBsToUse
 
-        -- Before traversing the install plan, preemptively find all packages that
-        -- will need to be downloaded and start downloading them.
-        asyncDownloadPackages
-          verbosity
-          withRepoCtx
-          installPlan
```

Comment:
Again fourmolu quizzical edits…

Reply:
Not really, due to bracket + do notation, I believe all indentation has to shift to the right.

Reply:
the only thing introduced in this pr is a `case`, not a `do`, right?

Reply:
Actually it's:

```diff

-        -- Before traversing the install plan, preemptively find all packages that
-        -- will need to be downloaded and start downloading them.
-        asyncDownloadPackages
+        bracket (pure mkJobControl) cleanupJobControl $ \jobControl -> do
+          -- Before traversing the install plan, preemptively find all packages that
+          -- will need to be downloaded and start downloading them.
+          asyncDownloadPackages
```